File: model/src/main/java/com/github/fburato/highwheelmodules/model/modules/ModuleGraphFactory.java
package com.github.fburato.highwheelmodules.model.modules;

import java.util.Optional;

public interface ModuleGraphFactory {
    MetricModuleGraph<ModuleDependency> buildMetricModuleGraph();

    ModuleGraph<EvidenceModuleDependency> buildEvidenceModuleGraph(ModuleGraph<TrackingModuleDependency> trackingGraph,
            Optional<Integer> evidenceLimit);

    ModuleGraph<TrackingModuleDependency> buildTrackingModuleGraph();
}

File: model/src/main/scala/com/github/fburato/highwheelmodules/model/modules/ModuleGraphFactory.scala
package com.github.fburato.highwheelmodules.model.modules

trait ModuleGraphFactory {

  def buildMetricModuleGraph: MetricModuleGraph[ModuleDependency]

  def buildEvidenceModuleGraph(
    moduleGraph: ModuleGraph[TrackingModuleDependency],
    evidenceLimit: Option[Int]
  ): ModuleGraph[EvidenceModuleDependency]

  def buildTrackingModuleGraph: ModuleGraph[TrackingModuleDependency]
}
